tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,401 @@
+//!         .unwrap()
+//! }
+//!
+//! let svc = ServiceBuilder::new()
+//!     // Use `handle_panic` to create the response.
+//!     .layer(CatchPanicLayer::custom(handle_panic))
+//!     .service_fn(handle);
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use bytes::Bytes;
+use futures_core::ready;
+use futures_util::future::{CatchUnwind, FutureExt};
+use http::{HeaderValue, Request, Response, StatusCode};
+use http_body::{combinators::UnsyncBoxBody, Body, Full};
+use pin_project_lite::pin_project;
+use std::{
+    any::Any,
+    future::Future,
+    panic::AssertUnwindSafe,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+use crate::BoxError;
+
+/// Layer that applies the [`CatchPanic`] middleware that catches panics and converts them into
+/// `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct CatchPanicLayer<T> {
+    panic_handler: T,
+}
+
+impl CatchPanicLayer<DefaultResponseForPanic> {
+    /// Create a new `CatchPanicLayer` with the default panic handler.
+    pub fn new() -> Self {
+        CatchPanicLayer {
+            panic_handler: DefaultResponseForPanic,
+        }
+    }
+}
+
+impl<T> CatchPanicLayer<T> {
+    /// Create a new `CatchPanicLayer` with a custom panic handler.
+    pub fn custom(panic_handler: T) -> Self
+    where
+        T: ResponseForPanic,
+    {
+        Self { panic_handler }
+    }
+}
+
+impl<T, S> Layer<S> for CatchPanicLayer<T>
+where
+    T: Clone,
+{
+    type Service = CatchPanic<S, T>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CatchPanic {
+            inner,
+            panic_handler: self.panic_handler.clone(),
+        }
+    }
+}
+
+/// Middleware that catches panics and converts them into `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy)]
+pub struct CatchPanic<S, T> {
+    inner: S,
+    panic_handler: T,
+}
+
+impl<S> CatchPanic<S, DefaultResponseForPanic> {
+    /// Create a new `CatchPanic` with the default panic handler.
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner,
+            panic_handler: DefaultResponseForPanic,
+        }
+    }
+}
+
+impl<S, T> CatchPanic<S, T> {
+    define_inner_service_accessors!();
+
+    /// Create a new `CatchPanic` with a custom panic handler.
+    pub fn custom(inner: S, panic_handler: T) -> Self
+    where
+        T: ResponseForPanic,
+    {
+        Self {
+            inner,
+            panic_handler,
+        }
+    }
+}
+
+impl<S, T, ReqBody, ResBody> Service<Request<ReqBody>> for CatchPanic<S, T>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+    ResBody::Error: Into<BoxError>,
+    T: ResponseForPanic + Clone,
+    T::ResponseBody: Body<Data = Bytes> + Send + 'static,
+    <T::ResponseBody as Body>::Error: Into<BoxError>,
```

Comment:
Do we actually need to include these send bounds?

Reply:
Yes. They're required because we call `http_body::Body::boxed_unsync` which requires the body to be `Send`. We call that on the body from inner service and from the callback that creates the panic response.